### ISSUE TITLE
Unexpected compilation error with switch

### ISSUE THREAD

**Issue by dlgoodchild** [Microsoft/TypeScript#15045]
**TypeScript Version:**  2.2.1

**Code**
Following is a reduced source that demonstrates the unexpected compilation error with a switch statement combined with the use of enums.
```ts
enum LoginStates {
	CallToAction,
	EnterInitials,
	SearchingUser
}

class Card {
	constructor( private config: LoginConfig ) {

	}

	private testSwichOnEnum() {
		let currentState:LoginStates = LoginStates.CallToAction;

		switch ( currentState ) {
			case LoginStates.CallToAction:
				break;

			case LoginStates.EnterInitials:
				this.stateElements.stateEnterInitials.addClass( 'front' );
				break;
		}
	}
}
```
Obviously in the above, I'm directly setting the currentState before I've fully fleshed out the class as a simple starting point. But I was surprised to see this error.

**Expected behavior:**
I would expect this to understand that currentState is a value and is then evaluated againsts each case.

**Actual behavior:**
error TS2678: Type 'LoginStates.EnterInitials' is not comparable to type 'LoginStates.CallToAction'.

**Comment by cevek** [comment]
Because actually for your example CFA(Control Flow Analyzer) detects that `currentState` has a value `LoginStates.CallToAction` and case `LoginStates.EnterInitials` is unnecessary here.

If you transform your example into function with `currentState` argument CFA would be know that `currentState` can possible any member of enum
```ts
function testSwichOnEnum(currentState:LoginStates) {
		switch ( currentState ) {
			case LoginStates.CallToAction:
				break;

			case LoginStates.EnterInitials:
				this.stateElements.stateEnterInitials.addClass( 'front' );
				break;
		}
	}
testSwichOnEnum(LoginStates.CallToAction);
```
